### ISSUE TITLE
Long-Term engine plans

### ISSUE THREAD

**Issue by IceReaper** [OpenRA/OpenRA#18448]
I did some thinking about making a documentation about OpenRA as an engine for game development.
However one of the biggest problems i see here are several key things still hardcoded to the core mods. People wanting to build something on openra defenitely have to bother with way to many required traits and definitions, and are at several key points limited to the engines posibility of not allowing to override specific stuff.

So i made a list of things i would love to see changed in OpenRA to make it a more general purpose game engine. Sure, it has a focus on RTS games, but games like fallout 2 or diablo should be totaly possible.

Ruleset:
The ruleset is in my eyes too strictly coded against the default implementation.
I would suggest to implement IRulesetLoader classes which are defined in mod.yaml.
Every IRulesetLoader then loads a specific type into the ruleset. Actors, Weapons, Voices, Notifications etc.
This allows two things:
1. It allows to implement different loaders to populate those dictionaries a different way
2. It allows to add new types. Maybe some games want "turrets" to be something by themself, and not an actor. Or in general items / equipment, without them being an actor.
3. It also allows to remove hardcoded stuff someone might not need. Example: Voices, Notifications, Tileset, etc.

WeaponInfo:
Just a minor thing here: TraitInfo has InstanceName for example. This pattern should be available across all yaml based things, like Warheads too. This is for example required to acces a specific Warhead node from code.

Map:
The whole default map implementation should be moved out of the engine, allowing mods to load any custom format they want, also defining any custom layer they want. So i suggest to remove most of the properties and any hard link to a tileset.
#18378 

Tileset:
The default tilesets do not support rotatable tiles or custom data on tilesets. So i would suggest to unhardcode tilesets, so everyone can implement their own variant.

Netcode:
I have several things here i would like to see or change myself:
1. First i would suggest to implement a generic lobby packet which mods can extend with custom data. This allows mods to share custom data on lobby-basis
2. I would suggest to do the same on client data. Allowing mods to add custom data to clients.
3. the whole logic of spawnpoints, teams, factions, bots etc should all go into the modular base. So mods may choose which mechanics they need to load or not. 
4. I would suggest adding an UDP socket for visual data. While the tcp layer still holds the main game communication, the udp socket could be used for features like direct-player-input. By default an actor may get a DirectControl trait, which overrides the visible location by the one provided by inputs (local) or data from the udp stream. This would allow people to implement wasd movement to units or such, without having a feelable input lag.

SaveGame:
Also i would like to see a complete refactor to make the current gamestate fully snapshotable. While this can replace the savegame function to a variant more update-friendly as the replay-approach, it also allows to implement features like re-joining or late-joining into games, by simply sending a full game snapshot over network. Some games might even work completely without a lobby at all. Hosting a game might throw you ingame, while other people may then late-join. It totaly depends on the game.

Scripting:
I would personaly love to see the lua stuff replaced by a script provider. Any mod then can choose whether a scripting engine is enabled, and if, also determine which. So specific mods may theoretically provide their own scripting language. One of those might be a complete flow-based visual scripting language.

Widgets:
First of all, i would love to see yoga implemented, as this allows for way more modular and responsive ui layouts.
#18151 

Platform:
I would prefer to have resizable windows available when yoga is implemented.
Also i would love to see a way to allow mods to specify a shader to use while drawing. Such a shader would come ofcourse with its own Vertex implementation and its own VertexBuffer layout. But that key requirement would allow mods to implement proper models, animations, lighting, effects etc.

Ofcourse this is more of a long-term roadmap, and every feature needs to be discusses, but in general its as said a long-term vision i would love to contribute to. So this issue here is meant for discussing above topics, working on a vision together, then moving individual ideas out to their own issues waiting for implementation.

**Comment by tovl** [comment]
I very much agree with the aim of evolving the OpenRA engine into a more flexible game engine. In fact, I wouldn't have joined this project in the first place if I didn't saw the potential here.

> Every IRulesetLoader then loads a specific type into the ruleset. Actors, Weapons, Voices, Notifications etc.

One of these is not like the others, though. Actors are already a very generic thing not tied to anything specific in the default mods. They are the entities in the entity-component model. In principle, they can be anything, not just units on the map, but also, say, the map itself could be an actor (not quite how it works right now, but we do have a world actor and player actors for instance).

In general I agree with you though. These things like weapons etc. somehow exist outside of the entity-component model and instead are their own hard-coded thing, while not being generic enough to justify that. I think most of these things could eventually be refactored to become either traits or actors. Instead of putting effort into providing hooks to add new distinct concepts like "turrets" to the mix, I would rather put that effort into improving the performance issues with trait lookups so more things can be reasonably modeled as actors/traits. I also think we might gradually want to move from the entity-component model to something more like an entity-component-system model, where the traits contain mostly just data and the logic is put into systems that are themselves modular.

> Widgets:

Speaking of widgets, I would like to see the viewport itself become a widget that can be placed and resized alongside other UI widgets. So, for instance, it would be possible to have the sidebar be more like in the original c&c where it does not overlap with the viewport or you could even have multiple viewports (like for instance in TTD with the pop-up windows).

> I would personaly love to see the lua stuff replaced by a script provider. Any mod then can choose whether a scripting engine is enabled, and if, also determine which. 

Multiple scripting engines seems a bit ambitious, without having any direct benefit. I do not see this become a high priority on anyone's list. Sure, I would love to be able to use Python instead of Lua, for instance, but, why go through the effort when Lua is mostly good enough?

As a general remark, while I agree with the overall aim of making the engine more modular and allow modders to add custom implementations of many pieces of engine logic if they want to, I think we should also keep in mind the limitations of that approach. The plain fact of the matter is, that most mod developers will be unable or unwilling to write engine code and I don't blame them one bit for that. Not every potential game developer is also a good low-level programmer. And the ones who are willing and able to do that, might as well just hard-fork the engine or write their own SDL wrappers or just use one of many more generic game engines that are already out there. Meanwhile, I've already seen modders get quite creative with combining and reusing existing traits for all kinds of unorthodox uses and I think that's great. 

So, my philosophy with regards to mod support is basically "Batteries included". Providing hooks for custom engine modules is nice for those who want to engage with that, but the meat is in providing a well-rounded and flexible set of common components/traits that can be freely combined and rearranged like lego bricks for a wide variety of uses. For me, that also means getting rid of overly specific traits and replacing them with more flexible ones that aren't tied to a specific mod. It also means allowing those generic traits to be defined multiple times on the same actor in different configurations, etc. In most cases this also happens to make the common library more maintainable, because there are, as it stands, many redundant traits that do mostly the same things with slight variations. This can lead to bitrot as a core engine feature gets updated and some traits are just forgotten.

TL;DR just like a game console is only as good as its game library, so too is a game engine only as good as the libraries available for it. So don't forget about `Mods.Common` and let's not fall into the trap of "OpenRA supports feature x because you can just make feature x from scratch".

**Comment by teinarss** [comment]
> In general I agree with you though. These things like weapons etc. somehow exist outside of the entity-component model and instead are their own hard-coded thing, while not being generic enough to justify that. I think most of these things could eventually be refactored to become either traits or actors. Instead of putting effort into providing hooks to add new distinct concepts like "turrets" to the mix, I would rather put that effort into improving the performance issues with trait lookups so more things can be reasonably modeled as actors/traits. I also think we might gradually want to move from the entity-component model to something more like an entity-component-system model, where the traits contain mostly just data and the logic is put into systems that are themselves modular.
> 

Refactoring to pure pure ECS would be great for the performance. Right now the data is scattered around in memory.

**Comment by dnqbob** [comment]
For RPG game there is a lot of engine which is the way better than OpenRA, for example, Unity3D. Input, physic logic, graphic, muti-platform and audio logic are all better than OpenRA. So if I try to make a RPG game which include some light weighted RTS I may still choose the Unity3D alike instead of OpenRA. However, OpenRA may need some very light weighted RPG system for mission only.

I think if OpenRA want to be a good game engine to develop on (and I also want it to), here is a key thing besides hard coding need dealing with first:

Art assets. Now we can only use SHP and VXL, while PNG can be used but poorly supported in rendering. If a game dev make a game by using OpenRA, he/she may face a welcome-to-court paper from EA since VXL and SHP are not open source and free for commercial use. So it will be vital for OpenRA support common RGBA and necessary rendering logic like player color and environment light.

**Comment by dnqbob** [comment]
Also, I want OpenRA finish its muti-language support.

**Comment by deleted-user-1** [comment]
> So this issue here is meant for discussing above topics, working on a vision together, then moving individual ideas out to their own issues waiting for implementation.

Just as a guidance for further comments.

**Comment by Porenutak** [comment]
As I read these long term goals I realize that no one care about Sound here. Which is sad because Sound is equal part of game experience as Graphic, game play, unit balance etc. So I write some ideas that I as a map maker and modder would like to see.
Disclaimer: I'm not coder, so don't know if they make any sense from code perspective. I write this only from content creator point of view.

Lets have separate yaml definition for SoundFX, and use reference to this definition in traits. Example:
Explosions property in effect Warhead, BounceSound in Bullet projectile or SoundFiles in AmbientSound trait. Instead of writing exact sound file, write mane from yaml definition where all properties for that SoundFX are set.

what can be in SoundFX yaml definition:
- various sounds file for one SoundFX. Example if I wanna definite BounceRock SFX it will include BounceRock01.wav, BounceRock02.wav, BounceRock03.wav sounds, and sequence they are play in(random, or as they appear)
- percentage change that sound is played.
Right now this can be done only in Create effect warhead, but with separate SoundFX definition u can have it in any trait that work with sound.

- Sound priority: if there are lot of explosions on screen u dont need to hear every debris bounce for example
- soundFX fade out/in with zoom. If player are fully zoom out some SFX dont need to by play(or play on low volume), and some may start to play.
- how should heard the sound: player only, allies only, enemy only
- and lastly (you probably head this): local sounds.

Thx for your attention

**Comment by IceReaper** [comment]
> As I read these long term goals I realize that no one care about Sound here. Which is sad because Sound is equal part of game experience as Graphic, game play, unit balance etc. So I write some ideas that as a map maker and modder I would like to see. Disclaimer: I'm not coder, so don't know if they make any sense from code perspective. I write this only from content creator point of view.

 Then do not forget:
- Random volume and random pitch support. This allows a single wav sound to sounds slightly differently every time
